pangeo_forge_recipes/executors/beam.py

```diff
@@ -0,0 +1,80 @@
+import itertools
+from dataclasses import dataclass
+from typing import Any, Iterable, List, Tuple, cast
+
+import apache_beam as beam
+
+from .base import Config, NoArgumentStageFunction, Pipeline, PipelineExecutor, Stage
+
+
+def _no_arg_stage(last: int, *, current: int, fun: NoArgumentStageFunction, config: Config) -> int:
+    """Execute a NoArgumentStageFunction, ensuring execution order."""
+    assert (last + 1) == current, f"stages are executing out of order! On step {current!r}."
+
+    fun(config=config)
+
+    return current
+
+
+def _no_op(arg, config=None) -> None:
+    pass
+
+
+@dataclass()
+class _SingleArgumentStage(beam.PTransform):
+    """Execute mappable stage in parallel."""
+
+    step: int
+    stage: Stage
+    config: Config
+
+    def prepare_stage(self, last: int) -> Iterable[Tuple[int, Any]]:
+        """Propagate current stage to Mappables for parallel execution."""
+        assert (last + 1) == self.step, f"stages are executing out of order! On step {self.step!r}."
+        return zip(itertools.repeat(self.step), cast(Iterable, self.stage.mappable))
+
+    def exec_stage(self, last: int, arg: Any) -> int:
+        """Execute stage function."""
+        assert last == self.step, f"stages are executing out of order! On step {self.step!r}."
+
+        self.stage.function(arg, config=self.config)  # type: ignore
+
+        return self.step
+
+    def post_validate(self, last: List[int]) -> int:
+        """Propagate step number for downstream stage validation."""
+        in_current_step = all((it == self.step for it in last))
+        assert in_current_step, f"stages are executing out of order! On step {self.step!r}."
+
+        return self.step
+
+    def expand(self, pcoll):
+        return (
+            pcoll
+            | "Prepare" >> beam.FlatMap(self.prepare_stage)
+            | beam.Reshuffle()
```

Comment:
I would suggest comparing this to just calling `beam.Create()` to create an in-memory collection of tasks. In my experience this can be a little more more efficient than always shuffling inputs, which can add a lot of overhead from anything being written to distributed disk.

If you use beam.Create, you can still ensure the ordering between stages by using a side input, like what I do here in xarray-beam: https://github.com/google/xarray-beam/blob/ebfdbf0275d226d86d4abb6cd47e46307c4f4036/xarray_beam/_src/zarr.py#L236

Reply:
Thanks for calling out that there will be overhead with writing to disk in the shuffle operation. However, I don't understand how `Create` could fit in here, and am not sure how the xarray-beam example can be applied. 

I notice that when I expand the pcoll with a `Create` step, I get an assertion error that the previous step is not a `PBegin`. This makes sense to me: Creates will only work when applied to the root of the pipeline. Maybe the side-input technique solves this problem, but I'm missing exactly how.

Reply:
Looking at the definition of `Create`, I notice that it also implements the FlatMap...Reshuffle pattern found in the above implementation:
```
  def expand(self, pbegin):
    assert isinstance(pbegin, pvalue.PBegin)
    serialized_values = [self._coder.encode(v) for v in self.values]
    reshuffle = self.reshuffle # this is true by default

    # Avoid the "redistributing" reshuffle for 0 and 1 element Creates.
    # These special cases are often used in building up more complex
    # transforms (e.g. Write).

    class MaybeReshuffle(PTransform):
      def expand(self, pcoll):
        if len(serialized_values) > 1 and reshuffle:
          from apache_beam.transforms.util import Reshuffle
          return pcoll | Reshuffle()
        else:
          return pcoll

    return (
        pbegin
        | Impulse()
        | FlatMap(lambda _: serialized_values).with_output_types(bytes)
        | MaybeReshuffle().with_output_types(bytes)
        | Map(self._coder.decode).with_output_types(self.get_output_type()))
```

Reply:
I don't understand enough about beam to know whether this conversation has been resolved.

Reply:
I'm convinced by @alxmrs's argument here.

We'll still have to see how well things work in practice, but theoretically I see no reason why this work any differently from what I was suggesting :+1: